### PR TITLE
[UPDATES] Algo for assigning proctors and candidates

### PR DESCRIPTION
**Pull request by maniSHarma7575**
**Problem Statement**

Write workflow for assigning the candidates and proctors to the room.

  - Candidate and proctor must be join to same room if somehow he left the room.
  - More than configurable no of candidates should not be visible to particular proctor.
  - If there are more than one proctor in room, candidate must be assigned to proctor in the first come first served basis.
  - If Proctor left the room somehow. He must joined the same room.

**Solution:**

**Functional Specifications:**

Candidate Side:

1. As soon as the candidate open the test window. All the checks [Microphone, Internet, Camera] will be performed.
2. Then after the Checks, the User will join the room.
3. As soon as he joins the room, his/her video will be render on his browser.
4. Whenever candidate reloads the page, the connection is lost. And he need to rejoin the room. In case of reload the Microphone, Internet and camera check will not be performed.
5. His/her message stream(video & audio) will be delivered to the proctor dashboard.

Proctor Side:

1. As soon as the proctor open the dashboard. First the loading screen will be shown to the proctor until he/she is connected to the room.
2. When the proctor is connected to the room. All the candidate in that room videos will be rendered on the proctoring dashboard.
3. Whenever the proctor, reload the page he will rejoin the room and rendering process happen again.

Will candidate be able to join the room without proctor?
- For now answer is yes, but we can have a feature to force the room to have at-least one proctor then only candidate will be able to join.

Can more than one proctor can participate in the room?
- Yes

**Technical Specifications:**

Algorithm for assigning the candidates and proctors:

1. How will candidates or proctors join?
    - As soon as the candidate page loads in the browser.
    - Candidates or proctors need to join the 100ms room.
    - For Joining the 100ms room required attributes:
         - userName
         - authToken
     - After joining the room, the candidate will be connected to a room.

7. How to generate authentication tokens?
    - A request will be sent from the browser to the application server for the authentication token.
    - Authentication token is a jwt token
   - To Create authentication token following attributes are required:
        - app_access_key -> will be fetched from the application credentials.
        - app_secret -> will be fetched from the application credentials.
        - user_id -> will be passed on the api request.
        - role -> will be passed on the api  request
        - room_id -> will re-use some old room_id or create new one.

8. How to generate a room_id for auth token?
    - For creating new room, we need to make api request to the 100ms server with attributes:
    - Management Token -> Is a jwt token created by the application.
         - Template Id -> For now from application credentials or can be created using the api request to 100ms.
         - S3 bucket -> from application credentials
         - Bucket location -> from application credentials
         - Folder prefix -> can be same as event_id
         - Credentials s3 -> from application credentials

9.  How will management token be generated?
     - Management token is a jwt token.
     - Management token attributes:
          - app_access_key
          - app_secret

10. When do we need to generate an authentication token?
     - First time Page Load -> First time when the candidate or proctor loads the browser then the authentication token will be fetched from server.
     - Page Reload -> If candidate reload the page then
Authentication token will be fetched from the local storage on the basis of event id.
If the fetched authentication token is expired then the authentication token will be again fetched from the server.

11. Algorithm for creating the room?

     - When this algorithm will be triggered: Whenever the user request a authentication token.
     - Creating Room depends on -> the no of candidates, no of proctors

     - Room creation algorithm:
 
Screenshot 2022-10-10 at 2 48 03 PM

### CODE REVIEW

File: app/assets/javascripts/100ms/hundred_ms.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    const authToken = getLocalStorageItem(authTokenKey);
+
+    if(props.userRole === 'proctor') setEventId(props);
+
+    if(authToken){
+      joinRoom({ userId: props.userId, authToken: authToken, userRole: props.userRole });
+    } else {
+      fetchAuthenticationToken(props).then((result) => {
+        if(result.success){
+          setLocalStorageWithExpiry(authTokenKey, result.authentication_token, 2 * 60 * 60);
```
Comment (nikhilbhatt) [inline app/assets/javascripts/100ms/hundred_ms.js:108]: what is this `2*60*60`? is it in minutes, seconds, milliseconds

File: app/assets/javascripts/100ms/proctor.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  
+    Object.keys(attrs).forEach((key) => {
+      newElement.setAttribute(key, attrs[key]);
+    })
+  
+    children.forEach((child) => {
+      newElement.append(child);
+    })
+  
+    return newElement
```
Comment (nikhilbhatt) [inline app/assets/javascripts/100ms/proctor.js:22]: `return newElement;`

File: app/assets/stylesheets/proctoring/video_player_100ms.css [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#proctoringVideos .videoContainer video {
+  width: 320px;
+  height: 240px;
+}
+.border-danger .status-online {
+  display: none;
+}
+.border-success .status-offline {
+  display: none;
+}
```
Comment (nikhilbhatt) [inline app/assets/stylesheets/proctoring/video_player_100ms.css:17]:
```
.border-danger .status-online, .border-success .status-offline {
  display: none;
}
```

Comment (nikhilbhatt) [reply app/assets/stylesheets/proctoring/video_player_100ms.css:17]: also, please stick to one naming convention. either camel case or kebab-case

File: app/controllers/proctoring/api/v1/authentication_controller.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      if room_id.blank?
+        room_id = generate_room({ event_id: event_id })['id']
+        add_room_to_the_cache(room_id, user_id, role, event_id)
+      end
+
+      authentication_params = { user_id: user_id, role: role, room_id: room_id }
+      authentication_token = generate_authentication_token(authentication_params)
+      render json: { success: true, authentication_token: authentication_token }
+    rescue StandardError => e
+      render json: { success: false, error: e.message }, status: :internal_error
```
Comment (nikhilbhatt) [inline app/controllers/proctoring/api/v1/authentication_controller.rb:28]: ```render json: { success: false, error: e.message }, status: :internal_server_error```

File: app/controllers/proctoring/application_controller.rb [hunk 5]
```diff
@@ -1,5 +1,17 @@
 module Proctoring
   class ApplicationController < ActionController::Base
     protect_from_forgery with: :exception
+
+    def authenticate_app_token
+      token_valid? request.headers['Token']
```
Comment (nikhilbhatt) [inline app/controllers/proctoring/application_controller.rb:6]: didn't understand this syntax here?

File: app/controllers/proctoring/application_controller.rb [hunk 6]
```diff
@@ -1,5 +1,17 @@
...
+    end
+
+    private
+
+    def token_valid?(token)
+      JWT.decode token, Proctoring.hundred_ms_auth_secret, true, { algorithm: 'HS256' }
+    rescue StandardError => _e
+      render json: { success: false, error: 'You are not the authorize person.' }, status: :unauthorized
```
Comment (nikhilbhatt) [inline app/controllers/proctoring/application_controller.rb:14]: `render json: { success: false, error: 'You are not an authorized person.' }, status: :unauthorized`

File: app/helpers/proctoring/hundred_ms_service_helper.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        candidates << { user_id: user_id, role: role }
+        rooms[room_id] = candidates
+        Rails.cache.write("100ms_room_details_#{event_id}", rooms, expires_in: 2.hours)
+        return room_id
+      end; nil
+    end
+
+    def add_room_to_the_cache(room_id, user_id, role, event_id)
+      candidates = []
+      candidates << { user_id: user_id, role: role }
```
Comment (nikhilbhatt) [inline app/helpers/proctoring/hundred_ms_service_helper.rb:48]:
```
candidates = [{
   user_id: user_id,
   role: role
}]
```

File: app/helpers/proctoring/tokens_helper.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+module Proctoring
+  module TokensHelper
+    def generate_authentication_token(params)
+      generate_app_token(
+        params[:room_id],
+        params[:user_id],
+        params[:role],
+        Proctoring.hundred_ms_app_access_key,
+        Proctoring.hundred_ms_app_secret
```
Comment (nikhilbhatt) [inline app/helpers/proctoring/tokens_helper.rb:13]: why a new method inside method? there is nothing else inside this other than calling a new method?

File: app/helpers/proctoring/tokens_helper.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        exp: exp.to_i
+      }
+      JWT.encode payload, Proctoring.hundred_ms_auth_secret, 'HS256'
+    end
+
+    private
+
+    def generate_app_token(room_id, user_id, role, app_access_key, app_secret)
+      now = Time.now
+      exp = now + 1.day
```
Comment (nikhilbhatt) [inline app/helpers/proctoring/tokens_helper.rb:43]: `now`? can we rename it? to current_time or something else? also `expiry_time` 

applies everywhere.

File: app/helpers/proctoring/tokens_helper.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        Proctoring.hundred_ms_app_access_key,
+        Proctoring.hundred_ms_app_secret
+      )
+    end
+
+    def generate_management_token
+      management_token(
+        Proctoring.hundred_ms_app_access_key,
+        Proctoring.hundred_ms_app_secret
+      )
```
Comment (nikhilbhatt) [inline app/helpers/proctoring/tokens_helper.rb:21]: same here.

File: app/assets/javascripts/100ms/hundred_ms.js (hunk 1 above)
Comment (maniSHarma7575) [reply app/assets/javascripts/100ms/hundred_ms.js:108]: It's in seconds

**Comment by maniSHarma7575** [comment]
Also closes #42 #43

**Pull request merged**